VPA Initializer, initial version.

bskiba opened the pull request:
VPA admission plugin equivalent implemented as an Initializer.
Changes declared requests of containers when applicable.

Moved mocks and test utils out of updater to reuse.

cc: @mwielgus @kgrygiel

mwielgus commented on vertical-pod-autoscaler/initializer/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+# Copyright 2016 The Kubernetes Authors. All rights reserved
```
2017

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+package main
+
+import (
+	"log"
```
organize importes

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+limitations under the License.
+*/
+
+package main
```
package initializer

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+)
+
+const (
+	vpaInitializerConfigName = "vpa-initializer-config.k8s.io"
```
Make it public. Add description what it is.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+type initializer struct {
+	client          kubeclient.Interface
+	vpaLister       apimock.VerticalPodAutoscalerLister
+	podListerSynced cache.InformerSynced
```
podLister

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	client          kubeclient.Interface
+	vpaLister       apimock.VerticalPodAutoscalerLister
+	podListerSynced cache.InformerSynced
+	informer        cache.SharedInformer
```
Is it needed?

bskiba replied:
I will be restructuring the initializer to be a more proper controller in the next PR, would you mind if I deal with this then?

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	config, err := i.registerer.Create(configuration)
+	if err != nil {
+		//TODO: If exists, return nil
+		glog.Infof("Error registering VPA Initializer Config: %v", err)
```
Errorf

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+		glog.Infof("Error registering VPA Initializer Config: %v", err)
+		return err
+	}
+	glog.V(4).Infof("Registered config: %+v", config)
```
This is probably quite important. V(0)/V(1)

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+)
+
+// Initializer initializes all pods matched by an existing vpa object
+// by assigning them recommended resource requests.
```
Expand it. Add information how it registers itself.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	glog.Infof("Starting VPA initializer.")
+	defer glog.Infof("Shutting down VPA initializer.")
+
+	i.register()
```
What if registration fails.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+func (i *initializer) unregister() {
+	i.registerer.Delete(vpaInitializerConfigName, &metav1.DeleteOptions{})
```
what if deletion fails?

mwielgus replied:
Does it even execute?

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+func (i *initializer) updateResourceRequests(obj interface{}) {
+	pod := obj.(*v1.Pod)
```
Add type checking, just to be safe.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+func shouldInitialize(pod *v1.Pod) bool {
+	return pod.ObjectMeta.Initializers != nil && len(pod.ObjectMeta.Initializers.Pending) > 0 && pod.ObjectMeta.Initializers.Pending[0].Name == vpaInitializerName
```
Confirm with @lavalamp if this is the correct check.

lavalamp replied:
looks right to me.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+		// TODO(bskiba): Right now this update always fails if we try to change resource requests
+		// as there are checks in place that prevent from changing requests of a created pod.
+		result, err := i.client.CoreV1().Pods(pod.Namespace).Update(initializedPod)
+		if err == nil {
```
Print error.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	initializedPod, err := i.getInitializedPod(pod)
+
+	if err == nil {
+		// TODO(bskiba): Right now this update always fails if we try to change resource requests
```
Please push this bug forward.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+
+}
+
+func getFailedPod(pod *v1.Pod) (*v1.Pod, error) {
```
markAsFailed

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+
+// getInitializedPod returns a pod in an initialized state, updating resource
+// requests if applicable.
+func (i *initializer) getInitializedPod(pod *v1.Pod) (*v1.Pod, error) {
```
initializePod?

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+		return nil, err
+	}
+	updatedPod := updatedPodCopy.(*v1.Pod)
+	glog.Infof("UPDATED POD: %+v", updatedPod)
```
Debug?

xiang90 replied:
why use capitalized letters here?

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+// requests if applicable.
+func (i *initializer) getInitializedPod(pod *v1.Pod) (*v1.Pod, error) {
+
+	glog.V(4).Infof("Updating requirements for pod: %v", pod.Name)
```
Add before/after or point me to the place where it is printed elsewhere.

bskiba replied:
This is printed for every updated container and resource in line 225/229

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+
+	vpaConfig := i.getMatchingVPA(pod)
+	if vpaConfig == nil {
+		glog.V(4).Infof("No matching vpa found.")
```
For what?

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+		}
+	}
+
+	glog.V(4).Infof("Recommendation for pod %v: %+v", pod.Name, recommendation)
```
More explicit debug please.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+			requested, exists := container.Resources.Requests[resource]
+			if exists {
+				// overwriting existing resource spec
+				glog.V(2).Infof("updating resources request for pod %v container %v resource %v old value %v new value %v",
```
Please define what is V(1), V(2) V(4) - i see that there are some inconsistencies in logging levels.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+// applyPolicy updates recommendation if recommended resources exceed limits defined in VPA resources policy
+func applyPolicy(recommendation *apimock.ContainerRecommendation, policy *apimock.ContainerPolicy) {
```
applyVpaPolicy

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	if policy != nil {
+		for _, container := range policy.Containers {
+			if containerName == container.Name {
+				return &container
```
&array[i]

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	}
+	for _, vpaConfig := range configs {
+		selector, err := labels.Parse(vpaConfig.Spec.Target.Selector)
+		glog.V(4).Infof("VPA: :%+v", vpaConfig)
```
Remove.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+		recommender: recommender.NewCachingRecommender(cacheTTl, apimock.NewRecommenderAPI()),
+	}
+
+	informer := newInformer(kubeClient, i)
```
single line.

mwielgus commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	return apimock.NewVpaLister(kubeClient)
+}
+
+func newInformer(kubeClient kubeclient.Interface, i *initializer) cache.SharedInformer {
```
you are passing a half-backed initiazliser, this is not a good patern. Please pass only the needed elements.

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+// Initializer initializes all pods matched by an existing vpa object
+// by assigning them recommended resource requests.
+type Initializer interface {
+	Run(stopCh <-chan struct{})
```
document public interface method?

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	}
+	failedPod, err := getFailedPod(pod)
+	if err == nil {
+		i.client.CoreV1().Pods(pod.Namespace).Update(failedPod)
```
handle error?

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	pod := obj.(*v1.Pod)
+
+	if !shouldInitialize(pod) {
+		glog.V(4).Infof("Not updating pod: %v", pod.Name)
```
this logging is not useful. logging why not updating is more useful.

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	}
+	initializedPod, err := i.getInitializedPod(pod)
+
+	if err == nil {
```
we probably should handle failed case first. it is more readable.

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+			recommendation = vpaConfig.Status.Recommendation
+		} else {
+			// no recommendation to apply
+			return updatedPod, nil
```
what if there is an error? we eat the error in this case.

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+// applyPolicy updates recommendation if recommended resources exceed limits defined in VPA resources policy
+func applyPolicy(recommendation *apimock.ContainerRecommendation, policy *apimock.ContainerPolicy) {
+	for resourceName, recommended := range recommendation.Resources {
+		if policy != nil {
```
too many indentations in this loop.

```
if policy == nil {
    continue
}

...

if !found {
    continue
}

...
}
```

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+func (i *initializer) getMatchingVPA(pod *v1.Pod) *apimock.VerticalPodAutoscaler {
+	configs, err := i.vpaLister.List()
+	if err != nil {
+		//cannot get vpa configs
```
probably not a good idea to eat all errors? wont list return other errors than not found?

xiang90 commented on vertical-pod-autoscaler/initializer/initializer_test.go:
```diff
@@ -0,0 +1,110 @@
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset/fake"
+)
+
+func TestUpdateResourceRequests(t *testing.T) {
```
please convert this to a table driven test.

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	registerer      admissionregistrationv1alpha1.InitializerConfigurationInterface
+}
+
+func (i *initializer) Run(stopCh <-chan struct{}) {
```
we should pass in context.Context to do cancellation.

bskiba replied:
I haven't seen this in Kubernetes codebase, can you point me to an example?

xiang90 commented on vertical-pod-autoscaler/initializer/initializer.go:
```diff
@@ -0,0 +1,347 @@
+	if err == nil {
+		i.client.CoreV1().Pods(pod.Namespace).Update(failedPod)
+	} else {
+		log.Fatalf("Unable to mark failed initialization for pod %v: %v", pod.Name, err)
```
why handle error==nil case first? usually we handle error case first when it is a fatal issue.

xiang90 commented on vertical-pod-autoscaler/initializer/main.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+var (
+	recommendationsCacheTtl = flag.Duration("recommendation-cache-ttl", 2*time.Minute,
```
recommendationsCacheTTL

xiang90 commented on vertical-pod-autoscaler/updater/updater.go:
```diff
@@ -21,16 +21,17 @@ import (
 	"k8s.io/client-go/tools/cache"
 	apiv1 "k8s.io/kubernetes/pkg/api/v1"
 
+	"time"
```
move standard import to top

bskiba commented:
Addressed most of the comments.
1. Organised imports
2. Restructured the code
3. Refactored tests
4. Revised logging
5. Added some comments
6. Fixed naming issues
7. Worked a bit on error reporting

I have not addressed all of the comments on the lifecycle/registering front. I will restructure the initializer with insights from [kubernetes initializer tutorial](https://github.com/kelseyhightower/kubernetes-initializer-tutorial) and [initializers best practices](https://github.com/kelseyhightower/kubernetes-initializer-tutorial/blob/master/docs/best-practices.md), but I would like to do it in the next PR if that's ok.

mwielgus commented on vertical-pod-autoscaler/initializer/core/initializer.go:
```diff
@@ -0,0 +1,377 @@
+
+func (initializer *initializer) doUpdatePod(pod *v1.Pod) error {
+	var err error
+	for i := 0; i < 5; i++ {
```
add sleep, make 5 constant

bskiba replied:
Done

mwielgus commented:
Almost there, 1 comment.

bskiba commented:
Awesome. Fixed comment, squashed commits.

mwielgus commented:
/lgtm
